File: app/src/main/java/com/krystianwsul/checkme/loaders/ShowNotificationGroupLoader.java
package com.krystianwsul.checkme.loaders;

import android.content.Context;
import android.support.annotation.NonNull;

import com.annimon.stream.Stream;
import com.krystianwsul.checkme.domainmodel.DomainFactory;
import com.krystianwsul.checkme.utils.InstanceKey;
import com.krystianwsul.checkme.utils.TaskKey;

import junit.framework.Assert;

import java.util.Set;

public class ShowNotificationGroupLoader extends DomainLoader<ShowNotificationGroupLoader.Data> {
    @NonNull
    private final Set<InstanceKey> mInstanceKeys;

    public ShowNotificationGroupLoader(@NonNull Context context, @NonNull Set<InstanceKey> instanceKeys) {
        super(context, needsFirebase(instanceKeys));

        Assert.assertTrue(!instanceKeys.isEmpty());

        mInstanceKeys = instanceKeys;
    }

    @NonNull
    private static FirebaseLevel needsFirebase(@NonNull Set<InstanceKey> instanceKeys) {
        return (Stream.of(instanceKeys)
                .map(InstanceKey::getType)
                .anyMatch(type -> type == TaskKey.Type.REMOTE) ? FirebaseLevel.NEED : FirebaseLevel.NOTHING);
    }

    @Override
    String getName() {
        return "ShowNotificationGroupLoader, instanceKeys: " + mInstanceKeys;
    }

    @Override
    public Data loadDomain(@NonNull DomainFactory domainFactory) {
        Assert.assertTrue(!mInstanceKeys.isEmpty());

        return domainFactory.getShowNotificationGroupData(getContext(), mInstanceKeys);
    }

    public static class Data extends DomainLoader.Data {
        @NonNull
        public final GroupListLoader.DataWrapper mDataWrapper;

        public Data(@NonNull GroupListLoader.DataWrapper dataWrapper) {
            mDataWrapper = dataWrapper;
        }

        @Override
        public int hashCode() {
            return mDataWrapper.hashCode();
        }

        @SuppressWarnings("RedundantIfStatement")
        @Override
        public boolean equals(Object object) {
            if (object == null)
                return false;

            if (object == this)
                return true;

            if (!(object instanceof Data))
                return false;

            Data data = (Data) object;

            if (!mDataWrapper.equals(data.mDataWrapper))
                return false;

            return true;
        }
    }
}

File: app/src/main/java/com/krystianwsul/checkme/loaders/ShowNotificationGroupLoader.kt
package com.krystianwsul.checkme.loaders

import android.content.Context

import com.krystianwsul.checkme.domainmodel.DomainFactory
import com.krystianwsul.checkme.gui.instances.tree.GroupListFragment
import com.krystianwsul.checkme.utils.InstanceKey
import com.krystianwsul.checkme.utils.TaskKey



class ShowNotificationGroupLoader(context: Context, private val instanceKeys: Set<InstanceKey>) : DomainLoader<ShowNotificationGroupLoader.Data>(context, needsFirebase(instanceKeys)) {

    companion object {

        private fun needsFirebase(instanceKeys: Set<InstanceKey>): DomainLoader.FirebaseLevel {
            return if (instanceKeys.any { it.type == TaskKey.Type.REMOTE })
                DomainLoader.FirebaseLevel.NEED
            else
                DomainLoader.FirebaseLevel.NOTHING
        }
    }

    init {
        check(!instanceKeys.isEmpty())
    }

    override val name = "ShowNotificationGroupLoader, instanceKeys: $instanceKeys"

    override fun loadDomain(domainFactory: DomainFactory): Data {
        check(!instanceKeys.isEmpty())

        return domainFactory.getShowNotificationGroupData(context, instanceKeys)
    }

    data class Data(val dataWrapper: GroupListFragment.DataWrapper) : DomainLoader.Data()
}
